File: test_prng.js
import { SeededRandom } from './prng.js';

async function testPrng(seed) {
    console.log(`JavaScript test with seed: ${seed}`);
    const rng = new SeededRandom(seed);
    // Wait for the seed to be initialized
    await rng.seedReady;

    // Create 10 buckets for values in [0,1)
    const numBuckets = 10;
    const buckets = new Array(numBuckets).fill(0);
    const numSamples = 1_000_000;
    const expectedPerBucket = numSamples / numBuckets;

    console.log(`Generating ${numSamples.toLocaleString()} random numbers...`);
    
    // Generate random numbers and count them in buckets
    for (let i = 0; i < numSamples; i++) {
        const value = rng.next();
        const bucketIndex = Math.floor(value * numBuckets);
        buckets[bucketIndex]++;
    }

    // Analyze the distribution
    console.log("\nDistribution analysis:");
    console.log("Bucket | Count    | Expected | Difference | % of Expected");
    console.log("-------|----------|----------|------------|--------------");
    
    let maxDeviation = 0;
    let maxDeviationBucket = 0;
    
    buckets.forEach((count, i) => {
        const difference = count - expectedPerBucket;
        const percentOfExpected = (count / expectedPerBucket * 100).toFixed(2);
        const deviation = Math.abs(difference);
        
        if (deviation > maxDeviation) {
            maxDeviation = deviation;
            maxDeviationBucket = i;
        }

        console.log(
            `${i.toString().padStart(5)} | ${count.toString().padStart(8)} | ${expectedPerBucket.toFixed(0).padStart(8)} | ${difference.toFixed(0).padStart(10)} | ${percentOfExpected.padStart(12)}%`
        );
    });

    // Calculate and display statistics
    const maxDeviationPercent = (maxDeviation / expectedPerBucket * 100).toFixed(2);
    console.log(`\nMaximum deviation: ${maxDeviation.toFixed(0)} (${maxDeviationPercent}%) in bucket ${maxDeviationBucket}`);
    
    // Chi-square test for uniformity
    const chiSquare = buckets.reduce((sum, count) => {
        const diff = count - expectedPerBucket;
        return sum + (diff * diff) / expectedPerBucket;
    }, 0);
    
    console.log(`Chi-square statistic: ${chiSquare.toFixed(2)}`);
    // For 9 degrees of freedom (10 buckets - 1), chi-square critical value at 0.05 is 16.92
    console.log("Chi-square critical value (0.05 significance): 16.92");
    console.log(`Distribution is ${chiSquare < 16.92 ? "uniform" : "not uniform"} at 0.05 significance level`);
}

// Get seed from command line or use default
const defaultSeed = "test_seed_123";
const seed = process.argv[2] || defaultSeed;

// Add usage information if no arguments or --help
if (process.argv.length === 2 || process.argv[2] === '--help') {
    console.log('Usage: node test_prng.js [seed]');
    console.log('  seed: Optional seed value for the PRNG (default: "test_seed_123")');
    console.log('  --help: Show this help message');
    if (process.argv.length === 2) {
        console.log('\nRunning with default seed...\n');
    } else {
        process.exit(0);
    }
}

testPrng(seed); 

File: test_prng.py
#!/usr/bin/env python3

import random
import sys
from typing import List

def test_prng(seed: str) -> None:
    print(f"Python test with seed: {seed}")
    rng = random.Random(seed)

    # Create 10 buckets for values in [0,1)
    num_buckets = 10
    buckets: List[int] = [0] * num_buckets
    num_samples = 1_000_000
    expected_per_bucket = num_samples / num_buckets

    print(f"Generating {num_samples:,} random numbers...")
    
    # Generate random numbers and count them in buckets
    for _ in range(num_samples):
        value = rng.random()
        bucket_index = int(value * num_buckets)
        buckets[bucket_index] += 1

    # Analyze the distribution
    print("\nDistribution analysis:")
    print("Bucket | Count    | Expected | Difference | % of Expected")
    print("-------|----------|----------|------------|--------------")
    
    max_deviation = 0
    max_deviation_bucket = 0
    
    for i, count in enumerate(buckets):
        difference = count - expected_per_bucket
        percent_of_expected = (count / expected_per_bucket * 100)
        deviation = abs(difference)
        
        if deviation > max_deviation:
            max_deviation = deviation
            max_deviation_bucket = i

        print(
            f"{i:5d} | {count:8d} | {expected_per_bucket:8.0f} | {difference:10.0f} | {percent_of_expected:12.2f}%"
        )

    # Calculate and display statistics
    max_deviation_percent = (max_deviation / expected_per_bucket * 100)
    print(f"\nMaximum deviation: {max_deviation:.0f} ({max_deviation_percent:.2f}%) in bucket {max_deviation_bucket}")
    
    # Chi-square test for uniformity
    chi_square = sum((count - expected_per_bucket) ** 2 / expected_per_bucket for count in buckets)
    
    print(f"Chi-square statistic: {chi_square:.2f}")
    # For 9 degrees of freedom (10 buckets - 1), chi-square critical value at 0.05 is 16.92
    print("Chi-square critical value (0.05 significance): 16.92")
    print(f"Distribution is {'uniform' if chi_square < 16.92 else 'not uniform'} at 0.05 significance level")

def print_usage() -> None:
    print('Usage: python test_prng.py [seed]')
    print('  seed: Optional seed value for the PRNG (default: "test_seed_123")')
    print('  --help: Show this help message')

if __name__ == "__main__":
    default_seed = "test_seed_123"
    
    # Handle command line arguments
    if len(sys.argv) == 1:
        print_usage()
        print('\nRunning with default seed...\n')
        seed = default_seed
    elif sys.argv[1] == '--help':
        print_usage()
        sys.exit(0)
    else:
        seed = sys.argv[1]
    
    test_prng(seed) 